### PR TITLE
[Impeller] Add a new API 'Context::WaitUntilCommandsCompleted' to ensure that commands are completed.

### PR DESCRIPTION
**Pull request by ColdPaleLight**
I'm working on https://github.com/flutter/flutter/issues/110829, and I found that Impeller doesn't have an API to ensure that all commands are completed.

Let's take the following code as an example. In the following piece of code, we will draw the `DisplayList` into a `Texture` in the Raster thread, and then blit the `Texture` into the `CPU Buffer` on the IO thread.  So we need to ensure that when we do the blit operation on the IO thread, the commands about the DisplayList drawn to the Texture have been completed in the GPU.
``` dart
final ui.PictureRecorder recorder = ui.PictureRecorder();
final ui.Canvas canvas = ui.Canvas(recorder);
canvas.drawRect(const Rect.fromLTWH(0.0, 0.0, 100, 100), Paint()..color = Colors.red);
final ui.Picture picture = recorder.endRecording();
ui.Image image = await picture.toImage(100, 100);
ByteData? byteData = await image.toByteData();
```

Another place where we need to use this new API is "blit texture to buffer". We need to ensure that the blit operation has been executed on the gpu before using the cpu buffer.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

### CODE REVIEW

File: impeller/renderer/context.h [hunk 1]
```diff
@@ -39,6 +39,8 @@ class Context : public std::enable_shared_from_this<Context> {
 
   virtual std::shared_ptr<CommandBuffer> CreateCommandBuffer() const = 0;
 
+  virtual bool WaitUntilCommandsCompleted() = 0;
```
Comment (dnfield) [inline impeller/renderer/context.h:42]: I don't think this belongs on `Context` - you want to know when the commands from a particular buffer are completed right?

For example, should this wait for all commands from compute to finish as well?

Comment (ColdPaleLight) [reply impeller/renderer/context.h:42]: I tried adding a parameter to the SubmitCommands method of CommandBuffer to solve the problem. But I found that in some scenarios, developers can't call SubmitCommands themselves, or even they can't find the Command they want to wait for.

for example
https://github.com/flutter/engine/blob/44649208521a19116330a1d0002d08f3361635a9/shell/common/snapshot_controller_impeller.cc#L66

Comment (dnfield) [reply impeller/renderer/context.h:42]: When you call `EntityPass::Render` from `AiksContext::Render` maybe there could be some parameter like a callback for when the texture is done?

Comment (dnfield) [reply impeller/renderer/context.h:42]: EntityPass::Render knows the command buffer, and the command buffer has `SubmitCommands(const CompletionCallback& callback)`.

Comment (ColdPaleLight) [reply impeller/renderer/context.h:42]: @dnfield Thanks you very much! I file a new PR https://github.com/flutter/engine/pull/37772, Any thoughts?

**Comment by ColdPaleLight** [comment]
Closing it since there is a better solution https://github.com/flutter/engine/pull/37772

**Pull request closed without merging**